### PR TITLE
feat(parser/renderer/types): Support inline and image icons

### LINKED ISSUE #587: Support inline icons
Eg:
```
icon:tags[] go libasciidoc
```

See https://asciidoctor.org/docs/user-manual/#inline-icons

### LINKED ISSUE #610: Support for image based icons
We should be able to support image based icons.

I'm fixing this as part of the fixes for inline icons.

### LINKED ISSUE #611: icon attribute support
There a bunch of attributes for icons that should be supported.  Size, rotation, flip, etc.  Also for image icons there are more like width, height.

We also want to support making these clickable links with the link attribute.

I'm addressing these all with the work to support inline icons.

### PR DESCRIPTION
**Pull request by gdamore**
This adds inline icon support, and significantly refactors the icon handling
to bring in support for image based icons and all of the icon attributes
supported by asciidoctor, bringing us to feature parity.

Additionally, we preserve some of the attributes from the font icon tags onto
images, allowing flip, rotate, and size styles to be applied to icons.
(This is an improvement over asciidoctor, and requires supporting CSS.)

One subtle difference when comparing output.  We emit source paths for images
without a leading "./".  That is because our use of path.Join() automatically
removes the redundant portion.

Fixes #587
Fixes #611
Fixes #610

### CODE REVIEW

File: pkg/parser/parser.peg [hunk 1]
```diff
@@ -388,7 +388,8 @@ TitleElement <- element:(Word
         / Space+ 
         / CrossReference
         / InlinePassthrough
-        / InlineImage 
+        / InlineIcon
+        / InlineImage
```
Comment (xcoulon) [inline pkg/parser/parser.peg:392]: could you add a few tests just to make sure that inline icons/images can be parsed in title elements?

File: pkg/parser/parser.peg [hunk 2]
```diff
@@ -717,6 +718,7 @@ LabeledListItemTermElement <- element:(Word
         / ConcealedIndexTerm
         / IndexTerm
         / InlinePassthrough
+        / InlineIcon
```
Comment (xcoulon) [inline pkg/parser/parser.peg:721]: could you add a few tests just to make sure that inline icons/images can be parsed in labeled list terms?

File: pkg/parser/parser.peg [hunk 3]
```diff
@@ -964,6 +967,7 @@ DoubleQuoteBoldTextElement <- Word
         / MonospaceText
         / SubscriptText
         / SuperscriptText
+        / InlineIcon
```
Comment (xcoulon) [inline pkg/parser/parser.peg:970]: could you add a few tests to verify that inline icons/images can be parsed in quoted text?

File: pkg/renderer/sgml/html5/icon.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	// the image to match rotate, flip, and size, allowing images to be manipulated with css style just
+	// like the icons can.  We also allow for an ID to be placed on the icon, as another enhancement.
+	inlineIconTmpl = `<span {{ if .ID }}id="{{ .ID }}" {{ end}}class="icon{{ if .Role }} {{ .Role }}{{ end }}">` +
+		`{{ if .Link }}<a class="image" href="{{ .Link }}"{{ if .Window }} target="{{ .Window }}"{{ end }}>{{ end }}` +
+		`{{ .Icon }}` +
+		`{{ if .Link }}</a>{{ end }}` +
+		`</span>`
+
+	iconImageTmpl = `<img src="{{ .Path }}"` +
+		`{{ if .Alt }} alt="{{ .Alt }}{{ end }}"` +
```
Comment (xcoulon) [inline pkg/renderer/sgml/html5/icon.go:15]: it's a good idea to use the raw strings to format the template! It makes it more readable 👍

File: pkg/types/types.go [hunk 5]
```diff
@@ -1047,7 +1047,7 @@ func NewParagraph(lines []interface{}, attributes interface{}) (Paragraph, error
 	if err != nil {
 		return Paragraph{}, errors.Wrapf(err, "failed to initialize a Paragraph element")
 	}
-	// log.Debugf("initializing a new paragraph with %d line(s) and %d attribute(s)", len(lines), len(attrs))
+	// log.Debugf("initializing a new Newparagraph with %d line(s) and %d attribute(s)", len(lines), len(attrs))
```
Comment (xcoulon) [inline pkg/types/types.go:1050]: this change could be reverted, but it's not a big deal since commented out anyways ;)

File: pkg/types/types.go [hunk 6]
```diff
@@ -1267,6 +1267,46 @@ func NewImageAttributes(alt, width, height interface{}, otherattrs []interface{}
...
+// ------------------------------------------
+// Icons
+// ------------------------------------------
+
+type Icon struct {
+	Class      string
+	Attributes Attributes
+}
+
+// NewImageAttributes returns a map of image attributes, some of which have implicit keys (`alt`, `width` and `height`)
```
Comment (xcoulon) [inline pkg/types/types.go:1279]:
```suggestion
// NewIconAttributes returns a map of icon attributes, some of which have implicit keys (`alt`, `width` and `height`)
```

File: pkg/types/types.go [hunk 7]
```diff
@@ -1267,6 +1267,46 @@ func NewImageAttributes(alt, width, height interface{}, otherattrs []interface{}
...
+					// mark custom_id flag to `true`
+					result = result.Set(AttrCustomID, true)
+				}
+			}
+		}
+	}
+	return result, nil
+}
+
+// NewInlineIcon initializes a new `InlineIcon`
```
Comment (xcoulon) [inline pkg/types/types.go:1302]:
```suggestion
// NewInlineIcon initializes a new `Icon`
```

File: pkg/types/types.go [hunk 8]
```diff
@@ -1267,6 +1267,46 @@ func NewImageAttributes(alt, width, height interface{}, otherattrs []interface{}
...
+func NewInlineIcon(class string, attributes Attributes) (Icon, error) {
```
Comment (xcoulon) [inline pkg/types/types.go:1303]: You could probably rename this func to `NewIcon`

File: pkg/renderer/sgml/html5/icon.go (hunk 4 above)
Comment (gdamore) [reply pkg/renderer/sgml/html5/icon.go:15]: Yeah, with this many if-blocks it was the only way I could do this and retain my sanity.

File: pkg/types/types.go (hunk 5 above)
Comment (gdamore) [reply pkg/types/types.go:1050]: lol.  I don't know that came in.

File: pkg/types/types.go (hunk 8 above)
Comment (gdamore) [reply pkg/types/types.go:1303]: Reasonable idea.  At one point I had a separate type for InlineIcon.  I guess I should clean this up too.

File: pkg/types/types.go (hunk 5 above)
Comment (gdamore) [reply pkg/types/types.go:1050]: I'm fixing this.

File: pkg/parser/parser.peg (hunk 2 above)
Comment (gdamore) [reply pkg/parser/parser.peg:721]: I'm doing this .. so they only work in labeled list *descriptions*.

The grammar does not permit any markup in  the *terms*.  This should probably be considered a defect, but not one that I'm going to fix right now.

Comment (gdamore) [reply pkg/parser/parser.peg:721]: (It's a preexisting defect.)

File: pkg/parser/parser.peg (hunk 1 above)
Comment (gdamore) [reply pkg/parser/parser.peg:392]: Yes.

File: pkg/parser/parser.peg (hunk 3 above)
Comment (gdamore) [reply pkg/parser/parser.peg:970]: Yes.

File: pkg/parser/parser.peg (hunk 1 above)
Comment (gdamore) [reply pkg/parser/parser.peg:392]: So it turns out that we only render plain here.  (Just like for images.)  It's possible that this decision could be revisited.  I think the ability to render images and icons inside a title could be useful.  What will happen now is you'll get the Alt text.

Comment (gdamore) [reply pkg/parser/parser.peg:392]: Ah, I guess that only happens for section 1.

Comment (gdamore) [reply pkg/parser/parser.peg:392]: Oh, wait.  I guess its for the table of contents only.  Lol.

File: pkg/parser/icon_test.go [hunk 9]
```diff
@@ -205,6 +204,118 @@ var _ = Describe("icons", func() {
...
+							},
+						},
+					},
+				}
+				Expect(ParseDraftDocument(source)).To(MatchDraftDocument(expected))
+
+			})
+
+			It("inline icon in quoted text", func() {
+				source := `an _italicized icon:warning[] message_`
```
Comment (xcoulon) [inline pkg/parser/icon_test.go:296]: sorry to insist, but could you add a test for each kind of quoted text? (bold, monospace, etc.)

**Comment by xcoulon** [comment]
> This adds inline icon support, and significantly refactors the icon handling
> to bring in support for image based icons and all of the icon attributes
> supported by asciidoctor, bringing us to feature parity.
> 
> Additionally, we preserve some of the attributes from the font icon tags onto
> images, allowing flip, rotate, and size styles to be applied to icons.
> (This is an improvement over asciidoctor, and requires supporting CSS.)
> 
> One subtle difference when comparing output. We emit source paths for images
> without a leading "./". That is because our use of path.Join() automatically
> removes the redundant portion.
> 
> Fixes #587
> Fixes #611
> Fixes #610

next time let's try to have a single PR per issue, so it's easier to review. But I imagine that you found out the bugs while working on the PR and probably wanted to address them immediatley?

**Comment by gdamore** [comment]
Yes, it would have been a pain to do this in separate PRs, the fixes were not easily separable.  I filed the tickets to log the issue, but splitting the work up not so easy.

File: pkg/parser/icon_test.go (hunk 9 above)
Comment (gdamore) [reply pkg/parser/icon_test.go:296]: Ugh.  Ok.  The parser tests are a royal pain.  I have come to really despise the test framework -- it is almost impossible to read the output when there is an error sometimes.

**Comment by xcoulon** [comment]
> Yes, it would have been a pain to do this in separate PRs, the fixes were not easily separable. I filed the tickets to log the issue, but splitting the work up not so easy.

sure, I understand the context here. It's just a good practice I want us to have here, to make PR focused and reviews simpler ;)

File: pkg/parser/icon_test.go (hunk 9 above)
Comment (xcoulon) [reply pkg/parser/icon_test.go:296]: you can try something like this:
```
ginkgo -focus "my test" pkg/parser -- -debug
```
with the `-debug` flag passed to the test, you'll have a dump of the actual result, so it's sometimes easier to compare with the expectation. Sometimes the diff is enough, but not always

**Comment by xcoulon** [comment]
thanks @gdamore 🙌

**Pull request merged**